File: share_memory/shared_memory.py
# Author: bdgoocci@gmail.com
# Description: refer to shared_memory.cpp 
import posix_ipc
import mmap
import os
import struct
import sys

class SharedMemory:

	int_mode = struct.Struct('i')
	
	def __init__(self, bf_size, shm_name, sem_post_name, sem_wait_name):
                self.buffer_size = bf_size - 8
		self.sem_post = posix_ipc.Semaphore(sem_post_name, posix_ipc.O_CREAT)
		self.sem_post_name = sem_post_name
		self.sem_wait = posix_ipc.Semaphore(sem_wait_name, posix_ipc.O_CREAT)
		self.sem_wait_name = sem_wait_name
                self.sem_wait.acquire()
		self.shm_obj = mmap.mmap(os.open(shm_name, os.O_RDWR), bf_size)
		self.shm_name = shm_name
		self.region_start = 0
		self.buffer_w_ptr = bf_size - 8
		self.buffer_r_ptr = bf_size - 4
                self.sem_post.release()

	def _def_(self):
		# override
		self.shm_obj.close()
		self.sem_post_name.close()
		self.sem_wait_name.close()

	def set_buffer_w_ptr(self, w_ptr_val):
		self.shm_obj.seek(self.buffer_w_ptr)
                if w_ptr_val >= self.buffer_size:
                    self.shm_obj.write(self.int_mode.pack(0))
                else:
                    self.shm_obj.write(self.int_mode.pack(w_ptr_val))

	def get_buffer_w_ptr(self):
		self.shm_obj.seek(self.buffer_w_ptr)
		return self.int_mode.unpack(self.shm_obj.read(4))[0]

	def set_buffer_r_ptr(self, r_ptr_val):
		self.shm_obj.seek(self.buffer_r_ptr)
                if r_ptr_val >= self.buffer_size:
                    self.shm_obj.write(self.int_mode.pack(0))
                else:
		    self.shm_obj.write(self.int_mode.pack(r_ptr_val))

	def get_buffer_r_ptr(self):
		self.shm_obj.seek(self.buffer_r_ptr)
		return self.int_mode.unpack(self.shm_obj.read(4))[0]

	def write(self, binstream):
		writing_bytes_progress = 0;
                
               # pdb.set_trace()
               
                while 1:
		        buffer_available = self.count_buffer_available()

		        if buffer_available < 4:
		 		self.sem_wait.acquire()
		 	else:
		 		self.write_binstream_size(len(binstream))
                                self.sem_post.acquire(0)
                                self.sem_post.release()
                                break

		while writing_bytes_progress < len(binstream):
		 	 buffer_available = self.count_buffer_available()
                         if buffer_available == 0:
		 	 	self.sem_wait.acquire()
		 	 	continue

		 	 writing_bytes_progress += self.write_binstream(binstream, buffer_available, writing_bytes_progress)

		 	 self.sem_post.acquire(0)
		 	 self.sem_post.release()


	def write_binstream_size(self, size):
		self.roll_write(self.int_mode.pack(size), 4)

	def write_binstream(self, binstr, buffer_avail, progress):
		if buffer_avail >= len(binstr) - progress:
			self.roll_write(binstr[progress:], len(binstr) - progress)
			return len(binstr) - progress
		else:
			self.roll_write(binstr[progress:], buffer_avail)
			return buffer_avail

	def count_buffer_available(self):
		buffer_w_ptr_val = self.get_buffer_w_ptr()
		buffer_r_ptr_val = self.get_buffer_r_ptr()
                
		if buffer_w_ptr_val == buffer_r_ptr_val:
			return self.buffer_size - 1
		elif buffer_w_ptr_val > buffer_r_ptr_val:
			return self.buffer_size - buffer_w_ptr_val + buffer_r_ptr_val - 1
		else:
			return buffer_r_ptr_val - buffer_w_ptr_val - 1

	def roll_write(self, bin, size):
               # pdb.set_trace()
		buffer_w_ptr_val = self.get_buffer_w_ptr()

		right_len = self.count_right_len()
                
		if right_len >= size:
			self.shm_obj.seek(buffer_w_ptr_val)
                        self.shm_obj.write(bin[:size])
                        self.set_buffer_w_ptr(buffer_w_ptr_val + size)
		else:
			self.shm_obj.seek(buffer_w_ptr_val)
			self.shm_obj.write(bin[:right_len])
			self.shm_obj.seek(0)
			self.shm_obj.write(bin[right_len:size])
                        self.set_buffer_w_ptr(size - right_len)

        
        def count_right_len(self):
                buffer_w_ptr_val = self.get_buffer_w_ptr()
                buffer_r_ptr_val = self.get_buffer_r_ptr()

                if buffer_r_ptr_val == 0:
                    return self.buffer_size - buffer_w_ptr_val - 1
                elif buffer_r_ptr_val <= buffer_w_ptr_val:
                    return self.buffer_size - buffer_w_ptr_val
                else:
                    return buffer_r_ptr_val - buffer_w_ptr_val - 1


        def read(self):
		msg = ''
                msg_len_bin = ''
               
                while 1:
		    msg_len_bin += self.roll_read(4 - len(msg_len_bin)) 
                    self.sem_post.acquire(0)
                    self.sem_post.release()
                    if len(msg_len_bin) < 4:
                        self.sem_wait.acquire()
                    else:
                        break
                
                #pdb.set_trace()
                msg_len = self.int_mode.unpack(msg_len_bin)[0]
                
		while len(msg) < msg_len:
                        msg += self.roll_read(msg_len - len(msg)) 
                 	self.sem_post.acquire(0)
			self.sem_post.release()

			if len(msg) < msg_len:
                            self.sem_wait.acquire()

                return msg


	def roll_read(self, size):
		r_ptr_val = self.get_buffer_r_ptr()
		w_ptr_val = self.get_buffer_w_ptr()
                bin = ''

                if r_ptr_val < w_ptr_val:
			max_read = w_ptr_val - r_ptr_val

			if max_read < size:
				self.shm_obj.seek(self.region_start + r_ptr_val)
				bin = self.shm_obj.read(max_read)
				self.set_buffer_r_ptr(r_ptr_val + max_read)
			else:
				self.shm_obj.seek(self.region_start + r_ptr_val)
				bin = self.shm_obj.read(size)         
				self.set_buffer_r_ptr(r_ptr_val + size)

		elif r_ptr_val > w_ptr_val:
			max_right_read = self.buffer_size - r_ptr_val
			max_left_read = w_ptr_val

			if size > max_right_read:
				self.shm_obj.seek(self.region_start + r_ptr_val)
				bin = self.shm_obj.read(max_right_read)
				surplus = size - max_right_read

				if surplus > max_left_read:
					self.shm_obj.seek(self.region_start)
					bin += self.shm_obj.read(max_left_read)
					self.set_buffer_r_ptr(max_left_read)
				else:
					self.shm_obj.seek(self.region_start)
					bin += self.shm_obj.read(surplus)
					self.set_buffer_r_ptr(surplus)

			else:
				self.shm_obj.seek(self.region_start + r_ptr_val)
				bin += self.shm_obj.read(size)
    				self.set_buffer_r_ptr(r_ptr_val + size)

		return bin 

File: share_memory/shared_memory.cpp
// Author: bdgoocci@gmail.com
// Description:  
//  assume 2 threads,
//  one thread called A is responsible for writing into the memory, 
//  the other thread called B reads the content which is written by thread A at the same time.
//  Semaphore is needed because there is a competition between threads.
//  If thread A and thread B are C++ threads, shared_memory.cpp is suitable.
//  If python thread is needed, please refer to shared_memory.py
#include "shared_memory.hpp"
#include "base/log.hpp"

namespace Husky {

	SharedMemory::SharedMemory(int bf_size, std::string shm_name, std::string sem_post_name, std::string sem_wait_name)
		: shm_obj(boost::interprocess::open_or_create, shm_name.c_str(), boost::interprocess::read_write),
		sem_post(boost::interprocess::open_or_create_t(), sem_post_name.c_str(), 0),
		sem_wait(boost::interprocess::open_or_create_t(), sem_wait_name.c_str(), 0) {
	    	buffer_size = bf_size - 8;
    		shm_obj.truncate(bf_size);
		region = boost::interprocess::mapped_region(shm_obj, boost::interprocess::read_write);
    		this->shm_name = shm_name;
    		this->sem_post_name = sem_post_name;
    		this->sem_wait_name = sem_wait_name;
    		region_start = reinterpret_cast<char*>(region.get_address());
    		buffer_w_ptr = reinterpret_cast<int*>(region_start + bf_size - 8);
    		buffer_r_ptr = reinterpret_cast<int*>(region_start + bf_size - 4);
    		*buffer_w_ptr = 0;
    		*buffer_r_ptr = 0;
            sem_post.post();
            sem_wait.wait();
	}

	SharedMemory::~SharedMemory() {
		boost::interprocess::shared_memory_object::remove(shm_name.c_str());
		boost::interprocess::named_semaphore::remove(sem_post_name.c_str());
		boost::interprocess::named_semaphore::remove(sem_wait_name.c_str());
	}

    void SharedMemory::write(std::string str) {  
        BinStream bin;
        bin.push_back_bytes(str.c_str(), str.size());
        write(bin);
    }

	void SharedMemory::write(BinStream& binstream) {
        // write params recording progress
        int writing_bytes_progress = 0;
        
        int buffer_available;
        while(true) {
            buffer_available = count_buffer_available();
            if (buffer_available < 4) {
                sem_wait.wait();
            } else {
                write_binstream_size(binstream.size());
                // log_msg("counter value before post writing size: " + std::to_string(sem_post.get_value()));
                sem_post.try_wait();
                sem_post.post();
                // log_msg("counter value after post writing size: " + std::to_string(sem_post.get_value()));
                break;
            }
        }
		
        while(writing_bytes_progress < binstream.size()) {
			// count buffer avaiable
            buffer_available = count_buffer_available();
			if (buffer_available == 0) {
				sem_wait.wait();
				continue;
			}

			writing_bytes_progress += write_binstream(binstream, buffer_available, writing_bytes_progress);
                                                
            // log_msg("counter value before post: " + std::to_string(sem_post.get_value()));
			sem_post.try_wait();
			sem_post.post();
            // log_msg("counter value after post: " + std::to_string(sem_post.get_value()));
		}

	}

	void SharedMemory::write_binstream_size(int size) {
		roll_write(reinterpret_cast<char*>(&size), 4);
	}

	int SharedMemory::write_binstream(BinStream& binstr, int buffer_avail, int progress) {
		if (buffer_avail >= binstr.size() - progress) {
			roll_write(binstr.get_remained_buffer() + progress, static_cast<int>(binstr.size()) - progress);
			return static_cast<int>(binstr.size()) - progress;
		} else {
			roll_write(binstr.get_remained_buffer() + progress, buffer_avail);
			return buffer_avail;
		}
	}

	int SharedMemory::count_buffer_available() {
        assert(*buffer_w_ptr >= 0);
        assert(*buffer_w_ptr < buffer_size);
        assert(*buffer_r_ptr >= 0);
        assert(*buffer_r_ptr < buffer_size);
		if (*buffer_w_ptr == *buffer_r_ptr) {
			return buffer_size - 1;
		} else if (*buffer_w_ptr > *buffer_r_ptr) {
			return buffer_size - *buffer_w_ptr + *buffer_r_ptr - 1;
		} else {
			return *buffer_r_ptr - *buffer_w_ptr - 1;
		}
	}

    int SharedMemory::count_right_len() {
        int buffer_w_ptr_val = *buffer_w_ptr;
        int buffer_r_ptr_val = *buffer_r_ptr;
        assert(*buffer_w_ptr >= 0);
        assert(*buffer_w_ptr < buffer_size);
        assert(*buffer_r_ptr >= 0);
        assert(*buffer_r_ptr < buffer_size);

        if (buffer_r_ptr_val == 0) {
            return buffer_size - buffer_w_ptr_val - 1;
        } else if (buffer_r_ptr_val <= buffer_w_ptr_val) {
            return buffer_size - buffer_w_ptr_val;
        } else {
            return buffer_r_ptr_val - buffer_w_ptr_val - 1;
        }
    }

	void SharedMemory::roll_write(const char* bin, int size) {
		int right_len = count_right_len();

        assert(*buffer_w_ptr >= 0);
        assert(*buffer_w_ptr < buffer_size);
        assert(*buffer_r_ptr >= 0);
        assert(*buffer_r_ptr < buffer_size);

		if (right_len >= size) {
			memcpy(region_start + *buffer_w_ptr, bin, size);
            assert(*buffer_w_ptr <= buffer_size);
            if (*buffer_w_ptr + size >= buffer_size) {
                assert(0 != *buffer_r_ptr);
                *buffer_w_ptr = 0;
            } else {
                assert(*buffer_w_ptr + size != *buffer_r_ptr);
                *buffer_w_ptr += size;
            }
		} else {
			memcpy(region_start + *buffer_w_ptr, bin, right_len);
			memcpy(region_start, bin + right_len, size - right_len);
			*buffer_w_ptr = size - right_len;
		}
               
        assert(*buffer_w_ptr >= 0);
        assert(*buffer_w_ptr < buffer_size);
        assert(*buffer_r_ptr >= 0);
        assert(*buffer_r_ptr < buffer_size);
	}

    std::string SharedMemory::read_string() {
        return read().to_string();
    }

	BinStream SharedMemory::read() {
		BinStream msg;
		BinStream msg_len_bin;
		int msg_len;
		int read_len = 0;
                
        while (true) {
            roll_read(msg_len_bin, 4 - msg_len_bin.size());
            sem_post.try_wait();
            sem_post.post();
            if (msg_len_bin.size() < 4) {
                sem_wait.post();    
            } else {
                break;
            }
        }
		msg_len = *reinterpret_cast<int*>(msg_len_bin.get_buffer());

		while(msg.size() < msg_len) {
			roll_read(msg, msg_len - msg.size());
            sem_post.try_wait();
			sem_post.post();
			if (msg.size() < msg_len) {
				sem_wait.wait();
			}
		}
		return msg;
	}

	void SharedMemory::roll_read(BinStream& bin, int size) {
		int r_pos = *buffer_r_ptr;
		int w_pos = *buffer_w_ptr;

        assert(*buffer_w_ptr >= 0);
        assert(*buffer_w_ptr < buffer_size);
        assert(*buffer_r_ptr >= 0);
        assert(*buffer_r_ptr < buffer_size);
		
        if (r_pos < w_pos) {
			int max_read = w_pos - r_pos;
			if (max_read < size) {
				bin.push_back_bytes(region_start + r_pos, max_read);
				*buffer_r_ptr = r_pos + max_read;
			} else {
				bin.push_back_bytes(region_start + r_pos, size);
				*buffer_r_ptr = r_pos + size;
			}
		} else if (r_pos > w_pos) {
			int max_right_read = buffer_size - r_pos;
			int max_left_read = w_pos;

			if (size > max_right_read) {
				// return right_read totally
				bin.push_back_bytes(region_start + r_pos, max_right_read);
				int surplus = size - max_right_read;

				if (surplus > max_left_read) {
					bin.push_back_bytes(region_start, max_left_read);
					*buffer_r_ptr = max_left_read;
				} else {
					bin.push_back_bytes(region_start, surplus);
					*buffer_r_ptr = surplus;
				}
			} else {
				bin.push_back_bytes(region_start + r_pos, size);
                if (r_pos + size >= buffer_size) {
                    *buffer_r_ptr = 0;
                } else {
				    *buffer_r_ptr = r_pos + size;
                }
			}
		}

        assert(*buffer_w_ptr >= 0);
        assert(*buffer_w_ptr < buffer_size);
        assert(*buffer_r_ptr >= 0);
        assert(*buffer_r_ptr < buffer_size);
	}	
}
